### PR TITLE
Tailing Facility Storage (TFS) POST method implementation.

### PR DESCRIPTION
**Pull request by areyeslo**
Jira Ticket: https://bcmines.atlassian.net/browse/MDS-3660

Front End changes to be able to edit a TSF record
Back End changes to be able to save the updates in DB.
Added tests for Front End and Back End.
Copied components in services/minespace-web/common/

### CODE REVIEW

File: services/core-api/app/api/mines/tailings/models/tailings.py [hunk 1]
```diff
@@ -54,9 +82,11 @@ def validate_tsf_name(self, key, mine_tailings_storage_facility_name):
         if len(mine_tailings_storage_facility_name) > 60:
             raise AssertionError('Mine name must not exceed 60 characters.')
         #no duplicate TSF names on the same mine
-        if (MineTailingsStorageFacility.query.filter_by(mine_guid=self.mine_guid).filter_by(
-                mine_tailings_storage_facility_name=mine_tailings_storage_facility_name).first() is
-                not None):
+        if (MineTailingsStorageFacility.query.filter_by(mine_guid=self.mine_guid).filter(
```
Comment (exzizt) [inline services/core-api/app/api/mines/tailings/models/tailings.py:82]: Good job fixing this bug

File: services/core-api/app/api/mines/tailings/resources/tailings.py [hunk 2]
```diff
@@ -122,3 +122,81 @@ def post(self, mine_guid):
...
+            raise NotFound('Mine not found.')
+
+        mine_tsf = MineTailingsStorageFacility.find_by_tsf_guid(mine_tailings_storage_facility_guid)
+
+        if not mine_tsf:
+            raise NotFound("Tailing Storage Facility not found")
+
+        data = self.parser.parse_args()
+
+        current_app.logger.info(f'Updating {mine_tsf} with {data}')
```
Comment (exzizt) [inline services/core-api/app/api/mines/tailings/resources/tailings.py]: Can we remove this logging? I don't think we need this in the logs

File: services/core-web/src/components/mine/Tailings/MineTailingsInfo.js [hunk 3]
```diff
@@ -76,6 +78,26 @@ export class MineTailingsInfo extends Component {
       .then(() => this.props.fetchMineReports(report.mine_guid, defaultParams.mineReportType));
   };
 
+  handleEditTailings = (values) => {
+    console.log("values:", values);
+    return this.props
```
Comment (mmumby) [inline services/core-web/src/components/mine/Tailings/MineTailingsInfo.js:80]: remove console.log()

File: services/core-api/tests/mines/tailings/resources/test_mine_tailings_resource.py [hunk 4]
```diff
@@ -40,40 +39,89 @@ def test_post_first_mine_tailings_storage_facility_by_mine_guid(test_client, db_
...
+    }
+
+    put_resp = test_client.put(
+        f'/mines/{uuid.uuid4()}/tailings/{uuid.uuid4()}',
+        data=data,
+        headers=auth_headers['full_auth_header'])
+    put_data = json.loads(put_resp.data.decode())
+
+    assert put_resp.status_code == 404
+    assert 'not found' in put_data['message']
```
Comment (exzizt) [inline services/core-api/tests/mines/tailings/resources/test_mine_tailings_resource.py]: No need to do this assertion, we do not do it in any other tests. It is ok just to check the status code

File: services/core-web/src/components/mine/Tailings/MineTailingsInfo.js [hunk 5]
```diff
@@ -76,6 +78,26 @@ export class MineTailingsInfo extends Component {
       .then(() => this.props.fetchMineReports(report.mine_guid, defaultParams.mineReportType));
   };
 
+  handleEditTailings = (values) => {
+    console.log("values:", values);
```
Comment (exzizt) [inline services/core-web/src/components/mine/Tailings/MineTailingsInfo.js]: Please remove console.log :)

File: services/core-web/src/components/mine/Tailings/MineTailingsInfo.js [hunk 6]
```diff
@@ -96,6 +118,26 @@ export class MineTailingsInfo extends Component {
...
+
+    const newRecord = { ...record, has_itrb: record.has_itrb.toString() };
+
+    event.preventDefault();
+    this.props.openModal({
+      props: {
+        initialValues: newRecord,
+        initialPartyValue,
+        onSubmit,
+        title: `Edit TSF for ${record.mine_tailings_storage_facility_name}`,
```
Comment (exzizt) [inline services/core-web/src/components/mine/Tailings/MineTailingsInfo.js]: I think this can just say "Edit ${record.mine_tailings_storage_facility_name}"

File: services/core-web/src/components/mine/Tailings/MineTailingsReportActions.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import React from "react";
```
Comment (mmumby) [inline services/core-web/src/components/mine/Tailings/MineTailingsReportActions.js]: I dont think you ended up using this file, this can be deleted.

**Review by exzizt — CHANGES_REQUESTED** [review]
Excellent work Arturo! Pretty much a perfect implementation in my opinion. Only thing I will ask is if you can break up the tailings.py resource file into two separate files, similar to how we have done with our other resources. Usually we have one called "name_resource.py" and "name_list_resource.py".

File: services/core-web/src/components/mine/Tailings/MineTailingsTable.js [hunk 8]
```diff
@@ -10,20 +10,26 @@ import * as Strings from "@common/constants/strings";
...
   isLoaded: PropTypes.bool.isRequired,
 };
 
 const defaultProps = {};
 
 export class MineTailingsTable extends Component {
-  transformRowData = (tailings) =>
-    tailings.map((tailing) => ({
+  transformRowData = (tailings) => {
+    return tailings.map((tailing) => ({
```
Comment (mmumby) [inline services/core-web/src/components/mine/Tailings/MineTailingsTable.js]: Can you revert the changes to this to make it an implicit return? 

```
transformRowData = (tailings) => (
this.props.... 
)
```

**Review by mmumby — DISMISSED** [review]
Great work, Arturo!

File: services/core-api/app/api/mines/tailings/resources/tailings.py (hunk 2 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/tailings/resources/tailings.py]: Done

File: services/core-web/src/components/mine/Tailings/MineTailingsInfo.js (hunk 3 above)
Comment (areyeslo) [reply services/core-web/src/components/mine/Tailings/MineTailingsInfo.js:80]: Done

File: services/core-api/app/api/mines/tailings/models/tailings.py [hunk 9]
```diff
@@ -20,10 +18,26 @@ class MineTailingsStorageFacility(AuditMixin, Base):
...
+
+    @hybrid_property
+    def engineer_of_record(self):
+        if self.engineer_of_records:
+            return self.engineer_of_records[0]
 
     def __repr__(self):
         return '<MineTailingsStorageFacility %r>' % self.mine_guid
 
+    def __str__(self):
```
Comment (exzizt) [inline services/core-api/app/api/mines/tailings/models/tailings.py]: Did you create this for debugging purposes? It could be removed as we don't really declare these for any of our classes. No big deal, though.

File: services/core-api/app/api/mines/tailings/resources/tailings.py [hunk 10]
```diff
@@ -48,67 +37,39 @@ class MineTailingsStorageFacilityListResource(Resource, UserMixin):
...
-            latitude=data['latitude'],
-            longitude=data['longitude'],
-            consequence_classification_status_code=data['consequence_classification_status_code'],
-            has_itrb=data['has_itrb'],
-            tsf_operating_status_code=data['tsf_operating_status_code'])
-        mine.mine_tailings_storage_facilities.append(mine_tsf)
+        if not mine_tsf:
+            raise NotFound("Tailing Storage Facility not found")
 
-        if is_mine_first_tsf:
```
Comment (exzizt) [inline services/core-api/app/api/mines/tailings/resources/tailings.py:92]: I'm guessing it was okayed to get rid of this logic?

File: services/core-web/common/constants/API.js [hunk 11]
```diff
@@ -20,7 +20,10 @@ export const MINE_NAME_LIST = (params = {}) => `/mines/search?${queryString.stri
 export const MINE_STATUS = "/mines/status";
 export const MINE_REGION = "/mines/region";
 export const MINE_COMPLIANCE_SUMMARY = (mine_guid) => `/mines/${mine_guid}/compliance/summary`;
-export const MINE_TSF = (mine_guid) => `/mines/${mine_guid}/tailings`;
+export const MINE_TSF = (mine_guid, mine_tailings_storage_facility_guid) =>
```
Comment (exzizt) [inline services/core-web/common/constants/API.js:24]: Can you please break this up into two separate constants? It would be more similar to how we've done others.

E.g.: MINE_TSF and MINE_TSFS

File: services/core-api/app/api/mines/tailings/models/tailings.py (hunk 9 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/tailings/models/tailings.py]: Yes, it was for sending to console the MineTailingStorageFacility object. It is removed :)

File: services/core-api/app/api/mines/tailings/resources/tailings.py (hunk 10 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/tailings/resources/tailings.py:92]: This is fixed.

File: services/core-web/common/constants/API.js (hunk 11 above)
Comment (areyeslo) [reply services/core-web/common/constants/API.js:24]: Decoupled the variables.

**Review by exzizt — DISMISSED** [review]
Thanks for making the requested changes and awesome work on your first big ticket! 😄 💯

**Review by mmumby — APPROVED** [review]
